File: problems/leetcode/0000/4/4.cpp
// 4. Median of Two Sorted Arrays
// https://leetcode.com/problems/median-of-two-sorted-arrays/description/

using namespace std;

#include <vector>

class Solution {
public:
  double findMedianSortedArrays(vector<int> &nums1, vector<int> &nums2) {
    int n1 = nums1.size(), n2 = nums2.size();
    vector<int> &smallerNums = n1 <= n2 ? nums1 : nums2,
                biggerNums = n1 <= n2 ? nums2 : nums1;
    int ns = smallerNums.size(), nb = biggerNums.size(), start = 0, end = ns;
    while (start <= end) {
      int pivotS = (start + end) / 2, pivotB = ((ns + nb + 1) / 2) - pivotS;

      int leftBoundaryS = pivotS == 0 ? INT_MIN : smallerNums[pivotS - 1];
      int rightBoundaryS = pivotS == ns ? INT_MAX : smallerNums[pivotS];
      int leftBoundaryB = pivotB == 0 ? INT_MIN : biggerNums[pivotB - 1];
      int rightBoundaryB = pivotB == nb ? INT_MAX : biggerNums[pivotB];

      if (leftBoundaryS <= rightBoundaryB && leftBoundaryB <= rightBoundaryS) {
        if ((ns + nb) % 2 == 0) {
          return (double)(((double)max(leftBoundaryS, leftBoundaryB) +
                           (double)min(rightBoundaryS, rightBoundaryB)) /
                          2);
        } else {
          return (double)(max(leftBoundaryS, leftBoundaryB));
        }
      } else if (leftBoundaryS > rightBoundaryB) {
        end = pivotS - 1;
      } else {
        start = pivotS + 1;
      }
    }

    return (double)0;
  }
};

File: problems/leetcode/0000/4/4.py
# 4. Median of Two Sorted Arrays
# https://leetcode.com/problems/median-of-two-sorted-arrays/description/

from typing import *

class Solution:
    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
        n1, n2 = len(nums1), len(nums2)
        if n1 > n2:
            return self.findMedianSortedArrays(nums2, nums1)

        start, end = 0, n1
        while start <= end:
            mid1 = (start + end) // 2
            mid2 = ((n1 + n2 + 1) // 2) - mid1

            left1 = nums1[mid1 - 1] if mid1 > 0 else float("-inf")
            right1 = nums1[mid1] if mid1 < n1 else float("inf")
            left2 = nums2[mid2 - 1] if mid2 > 0 else float("-inf")
            right2 = nums2[mid2] if mid2 < n2 else float("inf")

            if left1 <= right2 and left2 <= right1:
                return float(max(left1, left2) + min(right1, right2)) / 2 if (n1 + n2) % 2 == 0 else float(max(left1, left2))
            elif left1 > right2:
                end = mid1
            else:
                start = mid1 + 1
